File: lua/rootbeer/git.lua
--- @class git
local M = {}

local rb = require("rootbeer")
local tbl = require("rootbeer.tbl")

--- @class git.Config
--- @field path? string Where to write the gitconfig file. Defaults to `"~/.gitconfig"`.
--- @field user git.UserConfig User identity.
--- @field editor? string Default editor for commits (e.g. `"nvim"`).
--- @field pager? string Default pager for output (e.g. `"delta"`).
--- @field signing? git.SigningConfig Commit and tag signing. Sets `user.signingkey`, `gpg.format`, `commit.gpgSign`, and `tag.gpgSign`.
--- @field lfs? boolean Enable git-lfs filters (`filter.lfs` section).
--- @field pull_rebase? boolean Pull with rebase instead of merge.
--- @field merge_conflictstyle? string Merge conflict style (e.g. `"diff3"`).
--- @field ignores? string[] Global gitignore patterns. Written next to the gitconfig.
--- @field ignores_path? string Override path for the gitignore file. Defaults to `.gitignore` next to the gitconfig.
--- @field extra? table<string, table<string, string|boolean>> Additional gitconfig sections (e.g. `delta`, `interactive`). Merges into sections the other fields already wrote rather than replacing them, so `extra.core` keeps `excludesfile`. On a key conflict, `extra` wins.

--- @class git.UserConfig
--- @field name string Full name for commits.
--- @field email string Email address for commits.

--- @class git.SigningConfig
--- @field key string The signing key (e.g. an SSH public key).
--- @field format? string Signing format. Defaults to `"ssh"`.
--- @field allowed_signers? string[] Principals (usually emails) that `key` signs for. Writes an allowed signers file and points `gpg.ssh.allowedSignersFile` at it, which is what lets Git *verify* the signatures it creates. SSH format only.
--- @field allowed_signers_path? string Override path for the allowed signers file. Defaults to `"~/.ssh/allowed_signers"`.

--- Quotes a scalar for gitconfig format.
--- Strings are double-quoted; booleans and numbers are left bare.
--- Backslashes, quotes, and newlines are escaped so a value containing them
--- cannot terminate the string early and corrupt the surrounding section.
--- @param value string|number|boolean
--- @return string
local function quote(value)
	if type(value) ~= "string" then
		return tostring(value)
	end

	local escaped = value:gsub("\\", "\\\\"):gsub('"', '\\"'):gsub("\n", "\\n")
	return '"' .. escaped .. '"'
end

--- Recursively merges `src` into `dst`, with `src` winning on conflicts.
--- @param dst table
--- @param src table
local function merge_into(dst, src)
	for k, v in tbl.sorted_pairs(src) do
		if type(v) == "table" and type(dst[k]) == "table" then
			merge_into(dst[k], v)
		else
			dst[k] = v
		end
	end
end

--- Returns the directory portion of a path.
--- @param path string
--- @return string
local function dirname(path)
	return path:match("(.+)/") or "."
end

--- Emits a gitconfig-format string from a two-level table. Top-level keys
--- are sections; inner table values become `[section "subkey"]` blocks;
--- inner scalars become `\tkey = value` lines. Scalars are emitted
--- verbatim — callers should pre-quote strings via `quote()`.
--- @param sections table<string, table<string, string|table<string, string>>>
--- @return string
local function emit_gitconfig(sections)
	local parts = {}

	local function emit_block(header, body)
		if #parts > 0 then
			table.insert(parts, "")
		end
		table.insert(parts, header)
		for k, v in tbl.sorted_pairs(body) do
			if type(v) ~= "table" then
				table.insert(parts, "\t" .. k .. " = " .. tostring(v))
			end
		end
	end

	for section, values in tbl.sorted_pairs(sections) do
		-- subsections first: any inner key whose value is a table
		for k, v in tbl.sorted_pairs(values) do
			if type(v) == "table" then
				emit_block(string.format('[%s "%s"]', section, k), v)
			end
		end

		-- then the section's own scalars, if any
		local has_scalars = false
		for _, v in pairs(values) do
			if type(v) ~= "table" then
				has_scalars = true
				break
			end
		end
		if has_scalars then
			emit_block("[" .. section .. "]", values)
		end
	end

	return table.concat(parts, "\n") .. "\n"
end

--- Applies `git.Config` to the system. Writes a gitconfig file at `cfg.path`
--- and optionally a gitignore file next to it.
--- @param cfg git.Config
function M.config(cfg)
	local path = cfg.path or "~/.gitconfig"
	local ini = {}

	-- [user]
	ini.user = {
		name = cfg.user.name,
		email = cfg.user.email,
	}

	-- [core]
	local core = {}
	if cfg.editor then
		core.editor = cfg.editor
	end
	if cfg.pager then
		core.pager = cfg.pager
	end

	-- [commit], [tag], [gpg] via signing shortcut
	if cfg.signing then
		local format = cfg.signing.format or "ssh"
		ini.user.signingkey = cfg.signing.key
		ini.gpg = { format = format }
		ini.commit = { gpgSign = true }
		ini.tag = { gpgSign = true }

		-- Signing without this produces commits Git can create but not verify.
		if cfg.signing.allowed_signers and format == "ssh" then
			local signers_path = cfg.signing.allowed_signers_path
				or "~/.ssh/allowed_signers"
			local lines = {}
			for _, principal in ipairs(cfg.signing.allowed_signers) do
				lines[#lines + 1] = principal .. " " .. cfg.signing.key
			end

			rb.file(signers_path, table.concat(lines, "\n") .. "\n")
			ini.gpg.ssh = { allowedSignersFile = signers_path }
		end
	end

	-- ignores
	if cfg.ignores then
		local ignores_path = cfg.ignores_path
			or (dirname(path) .. "/.gitignore")
		core.excludesfile = ignores_path
		rb.file(ignores_path, table.concat(cfg.ignores, "\n") .. "\n")
	end

	if next(core) then
		ini.core = core
	end

	-- [pull]
	if cfg.pull_rebase ~= nil then
		ini.pull = { rebase = cfg.pull_rebase }
	end

	-- [merge]
	if cfg.merge_conflictstyle then
		ini.merge = { conflictstyle = cfg.merge_conflictstyle }
	end

	-- [filter "lfs"]
	if cfg.lfs then
		ini.filter = {
			lfs = {
				smudge = "git-lfs smudge -- %f",
				process = "git-lfs filter-process",
				required = true,
				clean = "git-lfs clean -- %f",
			},
		}
	end

	-- extra sections (delta, interactive, etc.). These merge into whatever the
	-- shortcuts above already wrote, so `extra.core` cannot drop `excludesfile`
	-- and `extra.tag` cannot drop `gpgSign`. On a key conflict, `extra` wins.
	if cfg.extra then
		for section, values in tbl.sorted_pairs(cfg.extra) do
			ini[section] = ini[section] or {}
			merge_into(ini[section], values)
		end
	end

	-- quote all string values for gitconfig format. Iteration order here
	-- doesn't matter since `emit_gitconfig` re-sorts; we just need to walk
	-- every entry to apply quoting.
	local quoted = {}
	for section, values in pairs(ini) do
		if type(values) == "table" then
			local q = {}
			for k, v in pairs(values) do
				if type(v) == "table" then
					local inner = {}
					for ik, iv in pairs(v) do
						inner[ik] = quote(iv)
					end
					q[k] = inner
				else
					q[k] = quote(v)
				end
			end
			quoted[section] = q
		end
	end

	rb.file(path, emit_gitconfig(quoted))
end

return M

File: crates/rootbeer-core/src/lua/tests/git.rs
//! Tests for `lua/rootbeer/git.lua`.

use crate::lua::test_support::run;
use crate::plan::Op;

fn writes(ops: &[Op]) -> Vec<(String, String)> {
    ops.iter()
        .filter_map(|op| match op {
            Op::WriteFile { path, source } => source
                .as_str()
                .map(|c| (path.display().to_string(), c.to_string())),
            _ => None,
        })
        .collect()
}

fn find<'a>(writes: &'a [(String, String)], suffix: &str) -> &'a str {
    &writes
        .iter()
        .find(|(p, _)| p.ends_with(suffix))
        .unwrap_or_else(|| panic!("no write for {suffix}; got: {writes:?}"))
        .1
}

#[test]
fn git_config_writes_user_section() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = "Alice", email = "alice@example.com" },
        })
        "#);
    let writes = writes(&ops);
    let cfg = find(&writes, "gitconfig");
    assert!(cfg.contains("[user]"));
    assert!(cfg.contains(r#"name = "Alice""#));
    assert!(cfg.contains(r#"email = "alice@example.com""#));
}

#[test]
fn git_config_signing_emits_commit_tag_gpg() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = "A", email = "a@b" },
            signing = { key = "ssh-ed25519 AAA" },
        })
        "#);
    let writes = writes(&ops);
    let cfg = find(&writes, "gitconfig");
    assert!(cfg.contains(r#"signingkey = "ssh-ed25519 AAA""#));
    assert!(cfg.contains("[gpg]"));
    assert!(cfg.contains(r#"format = "ssh""#));
    assert!(cfg.contains("[commit]"));
    assert!(cfg.contains("gpgSign = true"));
    assert!(cfg.contains("[tag]"));
}

#[test]
fn git_config_writes_ignores_file_alongside_config() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = "A", email = "a@b" },
            ignores = { ".DS_Store", "*.swp" },
        })
        "#);
    let writes = writes(&ops);

    let ignore = find(&writes, ".gitignore");
    assert!(ignore.contains(".DS_Store"));
    assert!(ignore.contains("*.swp"));

    let cfg = find(&writes, "gitconfig");
    assert!(cfg.contains("excludesfile = "));
    assert!(cfg.contains("/tmp/rb-test/.gitignore"));
}

#[test]
fn git_config_lfs_emits_filter_lfs_subsection() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = "A", email = "a@b" },
            lfs = true,
        })
        "#);
    let writes = writes(&ops);
    let cfg = find(&writes, "gitconfig");
    assert!(cfg.contains(r#"[filter "lfs"]"#));
    assert!(cfg.contains("clean ="));
    assert!(cfg.contains("smudge ="));
}

#[test]
fn git_config_extra_merges_into_shortcut_sections() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = "A", email = "a@b" },
            editor = "nvim",
            signing = { key = "ssh-ed25519 AAA" },
            ignores = { ".DS_Store" },
            extra = {
                core = { fsmonitor = true },
                tag = { sort = "-version:refname" },
                gpg = { ssh = { allowedSignersFile = "~/.ssh/allowed_signers" } },
            },
        })
        "#);
    let writes = writes(&ops);
    let cfg = find(&writes, "gitconfig");

    // extra.core must not drop what `editor` and `ignores` wrote.
    assert!(cfg.contains(r#"editor = "nvim""#));
    assert!(cfg.contains("excludesfile = "));
    assert!(cfg.contains("fsmonitor = true"));

    // extra.tag must not drop the signing shortcut's gpgSign.
    assert!(cfg.contains("gpgSign = true"));
    assert!(cfg.contains(r#"sort = "-version:refname""#));

    // extra.gpg adds a subsection without dropping the scalar format.
    assert!(cfg.contains(r#"format = "ssh""#));
    assert!(cfg.contains(r#"[gpg "ssh"]"#));
}

#[test]
fn git_config_extra_wins_on_conflict() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = "A", email = "a@b" },
            pull_rebase = true,
            extra = { pull = { rebase = false } },
        })
        "#);
    let writes = writes(&ops);
    let cfg = find(&writes, "gitconfig");
    assert!(cfg.contains("rebase = false"));
    assert!(!cfg.contains("rebase = true"));
}

#[test]
fn git_config_quotes_are_escaped() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = [[A "Quoted" B]], email = "a@b" },
            extra = { alias = { win = [[!f() { echo C:\temp; }; f]] } },
        })
        "#);
    let writes = writes(&ops);
    let cfg = find(&writes, "gitconfig");
    assert!(cfg.contains(r#"name = "A \"Quoted\" B""#));
    assert!(cfg.contains(r#"C:\\temp"#));
}

#[test]
fn git_config_allowed_signers_writes_file_and_wires_config() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = "A", email = "a@b" },
            signing = {
                key = "ssh-ed25519 AAA",
                allowed_signers = { "a@b", "a@work" },
                allowed_signers_path = "/tmp/rb-test/allowed_signers",
            },
        })
        "#);
    let writes = writes(&ops);

    let signers = find(&writes, "allowed_signers");
    assert_eq!(signers, "a@b ssh-ed25519 AAA\na@work ssh-ed25519 AAA\n");

    let cfg = find(&writes, "gitconfig");
    assert!(cfg.contains(r#"[gpg "ssh"]"#));
    assert!(cfg.contains(r#"allowedSignersFile = "/tmp/rb-test/allowed_signers""#));
}

#[test]
fn git_config_allowed_signers_skipped_for_non_ssh_format() {
    let ops = run(r#"
        local git = require("rootbeer.git")
        git.config({
            path = "/tmp/rb-test/gitconfig",
            user = { name = "A", email = "a@b" },
            signing = {
                key = "ABCD1234",
                format = "openpgp",
                allowed_signers = { "a@b" },
            },
        })
        "#);
    let writes = writes(&ops);
    assert!(!writes.iter().any(|(p, _)| p.ends_with("allowed_signers")));

    let cfg = find(&writes, "gitconfig");
    assert!(cfg.contains(r#"format = "openpgp""#));
    assert!(!cfg.contains("allowedSignersFile"));
}
